activestorage/s3.py

```diff
@@ -0,0 +1,128 @@
+    api_operation = "sum" if operation == "mean" else operation or "select"
+    url = f'{server}/v1/{api_operation}/'
+    response = request(url, username, password, request_data)
+
+    if response.ok:
+        # FIXME: Return count from mean
+        result = decode_result(response)
+        if operation == "mean":
+            count = reduce_chunk(server, username, password, source, bucket, object, offset, size, compression, filters, missing, dtype, shape, order, chunk_selection, "count")[0]
+        else:
+            count = None
+        return result, count
+    else:
+        decode_and_raise_error(response)
+
+
+def encode_selection(selection):
+    """Encode a chunk selection in a JSON-compatible format."""
+    def encode_slice(s):
+        if isinstance(s, slice):
+            return [s.start, s.stop, s.step]
+        else:
+            # Integer - select single value
+            return [s, s + 1, 1]
+
+    return [encode_slice(s) for s in selection]
+
+
+def build_request_data(source: str, bucket: str, object: str, offset: int,
+                       size: int, compression, filters, missing, dtype, shape,
+                       order, selection) -> dict:
+    """Build request data for S3 active storage API."""
+    # TODO: compression, filters, missing
+    request_data = {
+        'source': source,
+        'bucket': bucket,
+        'object': object,
+        'dtype': dtype.name,
+        'offset': offset,
+        'size': size,
+        'order': order,
+    }
+    if shape:
+        request_data["shape"] = shape
+    if selection:
+        request_data["selection"] = encode_selection(selection)
+    return {k: v for k, v in request_data.items() if v is not None}
+
+
+def request(url: str, username: str, password: str, request_data: dict):
+    """Make a request to an S3 active storage API."""
+    response = requests.post(
+        url,
+        json=request_data,
+        auth=(username, password)
+    )
+    return response
+
+
+def decode_result(response):
+    """Decode a successful response, return as a numpy array or scalar."""
+    dtype = response.headers['x-activestorage-dtype']
+    shape = json.loads(response.headers['x-activestorage-shape'])
+    result = np.frombuffer(response.content, dtype=dtype)
+    result = result.reshape(shape)
+    return result
+
+
+class S3ActiveStorageError(Exception):
+    """Exception for S3 Active Storage failures."""
+
+    def __init__(self, status_code, error):
+        super(S3ActiveStorageError, self).__init__(f"S3 Active Storage error: HTTP {status_code}: {error}")
+
+
+def decode_and_raise_error(response):
+    """Decode an error response and raise S3ActiveStorageError."""
+    try:
+        error = json.dumps(response.json())
+        raise requests.exceptions.JSONDecodeError
+        raise S3ActiveStorageError(response.status_code, error)
```

Comment:
I am fairly sure that once the first exception is raised (the one from requests) then the S3Active is no longer picked up and raised, so you'd have to decide on a raised exception of the two

Reply:
Correct, this is a coding error.